src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php

```diff
@@ -76,6 +76,10 @@ public function setUser($user)
             throw new \InvalidArgumentException('$user must be an instanceof UserInterface, an object implementing a __toString method, or a primitive string.');
         }
 
+        if (!$user instanceof UserInterface) {
+            @trigger_error(sprintf('Storing a user not implementing the %s is deprecated since Symfony 5.1.', UserInterface::class));
```

Comment:
Wouldn't this cause deprecations on anonymous authentications? The AnonymousToken also calls `setUser()` As proposed back in #3697, perhaps an AnonymousUser would make sense? This could be a stale object always returning the same information.

https://github.com/symfony/symfony/blob/e60a876201b5b306d0c81a24d9a3db997192079c/src/Symfony/Component/Security/Core/Authentication/Token/AnonymousToken.php#L33

Reply:
Yes, for me pushed code should just be the starting point. Actually, I was a bit surprised that our code did not trigger any deprecations. Might be an indicator that we do miss some tests (the anonymous user is one of the examples where I did expect the new deprecation to be triggered).

Reply:
my proposal applied to `ControllerTrait::getUser()`, to keep it consistent with `Security::getUser()`.

no need to patch `TokenInterface` IMHO, as it's low-level API, unless we decide to refactor everything now..? :joy:

Reply:
I think this change is good to make the usage consistent for now. At a later stage, I hope we can get rid of the UserInterface inside the token altogether.

Reply:
@ro0NL IMO it would be too confusing if the controller method behaved differently than the token storage.

Reply:
IMHO it's confusing for `Security::getUser` to behave different already.

They way i see it is, `Security/ControllerTrait::getUser()` is high-level API, designed around `UserInterface|null`.

Whereas `TokenInterface::getUser()`, the low-level API, is designed around `object|string|null`.

So the high level API is a subset.